### PR TITLE
Misc cleanup in PushContext

### PR DESCRIPTION
**Pull request by nmittler**
**Please provide a description of this PR:**

### CODE REVIEW

**Comment by nmittler** [comment]
FYI @stevenctl @howardjohn took a look through the [benchmark results] and it appears there is no difference.

**Review by howardjohn — COMMENTED** [review]
Will update PR comments after some offline discussion

File: pilot/pkg/model/serviceindex.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+	return &serviceIndexImpl{
+		public:                      []*Service{},
+		privateByNamespace:          map[string][]*Service{},
+		exportedToNamespace:         map[string][]*Service{},
+		hostnameAndNamespace:        map[host.Name]namespaceIndexImpl{},
+		serviceInstanceIndexesByKey: map[string]serviceInstanceIndexImpl{},
+	}
+}
+
+func SetServiceIndexHostsAndNamespacesForTesting(si ServiceIndex, hostnameAndNamespace map[host.Name]map[string]*Service) {
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/serviceindex.go:208]: Please move this to test files

File: pilot/pkg/model/serviceindex.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+type serviceIndexImpl struct {
+	// privateByNamespace are services that can reachable within the same namespace, with exportTo "."
+	privateByNamespace map[string][]*Service
+	// public are services reachable within the mesh with exportTo "*"
+	public []*Service
+	// exportedToNamespace are services that were made visible to this namespace
+	// by an exportTo explicitly specifying this namespace.
+	exportedToNamespace map[string][]*Service
+
+	hostnameAndNamespace map[host.Name]namespaceIndexImpl
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/serviceindex.go:145]: add comments

File: pilot/pkg/model/serviceindex.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+func (si serviceInstanceIndexImpl) Instances(port int) []*ServiceInstance {
+	return si[port]
+}
+
+func (si serviceInstanceIndexImpl) ForEach(visitor func(*ServiceInstance) bool) {
+	for _, instances := range si {
+		for _, inst := range instances {
+			if !visitor(inst) {
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/serviceindex.go:123]: This is confusing purely looking at this function

File: pilot/pkg/model/serviceindex.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+	if out == nil {
+		return emptyServiceInstanceIndex{}
+	}
+	return out
+}
+
+func (si *serviceIndexImpl) NamespacesForHost(h host.Name) NamespaceIndex {
+	ni := si.hostnameAndNamespace[h]
+	if ni == nil {
+		return emptyNamespaceIndex{}
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/serviceindex.go:164]: why  not  return nil

File: pilot/pkg/model/serviceindex.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+func SetServiceIndexHostsAndNamespacesForTesting(si ServiceIndex, hostnameAndNamespace map[host.Name]map[string]*Service) {
+	castSI := si.(*serviceIndexImpl)
+
+	for k, v := range hostnameAndNamespace {
+		castSI.hostnameAndNamespace[k] = v
+	}
+}
+
+func SetServiceIndexServiceForHostAndNamespaceForTesting(si ServiceIndex, h host.Name, ns string, s *Service) {
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/serviceindex.go:216]: same for these testing functions, should go to test files

**Comment by nmittler** [comment]
Since the scope on this PR changes and it's not really needed anymore, I'm just going to close this.

**Pull request closed without merging**